File: projects/java-sdk/src/main/java/com/example/java_sdk/EventsController.java
package com.example.java_sdk;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fingerprint.api.FingerprintApi;
import com.fingerprint.model.EventsGetResponse;
import com.fingerprint.model.EventsUpdateRequest;
import com.fingerprint.model.SearchEventsResponse;
import com.fingerprint.sdk.ApiClient;
import com.fingerprint.sdk.ApiException;
import com.fingerprint.sdk.ApiResponse;
import com.fingerprint.sdk.Configuration;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
public class EventsController {
    private final ObjectMapper objectMapper;

    public EventsController(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @GetMapping("/searchEvents")
    public ResponseEntity<MusicianResponse> searchEvents(
            @RequestParam(required = false) String apiKey,
            @RequestParam(required = false) String region,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false) String paginationKey,
            @RequestParam(required = false) String visitorId,
            @RequestParam(required = false) String bot,
            @RequestParam(required = false) String ipAddress,
            @RequestParam(required = false) String linkedId,
            @RequestParam(required = false) Long start,
            @RequestParam(required = false) Long end,
            @RequestParam(required = false) Boolean reverse,
            @RequestParam(required = false) Boolean suspect,
            @RequestParam(required = false) Boolean vpn,
            @RequestParam(required = false) Boolean virtualMachine,
            @RequestParam(required = false) Boolean tampering,
            @RequestParam(required = false) Boolean antiDetectBrowser,
            @RequestParam(required = false) Boolean incognito,
            @RequestParam(required = false) Boolean privacySettings,
            @RequestParam(required = false) Boolean jailbroken,
            @RequestParam(required = false) Boolean frida,
            @RequestParam(required = false) Boolean factoryReset,
            @RequestParam(required = false) Boolean clonedApp,
            @RequestParam(required = false) Boolean emulator,
            @RequestParam(required = false) Boolean rootApps,
            @RequestParam(required = false) Float minSuspectScore,
            @RequestParam(required = false) Boolean ipBlocklist,
            @RequestParam(required = false) Boolean datacenter
    ) {
        ApiClient client = Configuration.getDefaultApiClient(apiKey, Utils.getRegion(region));
        FingerprintApi api = new FingerprintApi(client);
        try {
            final ApiResponse<SearchEventsResponse> apiResponse =
                    api.searchEventsWithHttpInfo(limit, new FingerprintApi.SearchEventsOptionalParams()
                            .setPaginationKey(paginationKey)
                            .setVisitorId(visitorId)
                            .setBot(bot)
                            .setIpAddress(ipAddress)
                            .setLinkedId(linkedId)
                            .setStart(start)
                            .setEnd(end)
                            .setReverse(reverse)
                            .setSuspect(suspect)
                            .setVpn(vpn)
                            .setVirtualMachine(virtualMachine)
                            .setTampering(tampering)
                            .setAntiDetectBrowser(antiDetectBrowser)
                            .setIncognito(incognito)
                            .setPrivacySettings(privacySettings)
                            .setJailbroken(jailbroken)
                            .setFrida(frida)
                            .setFactoryReset(factoryReset)
                            .setClonedApp(clonedApp)
                            .setEmulator(emulator)
                            .setRootApps(rootApps)
                            .setMinSuspectScore(minSuspectScore)
                            .setIpBlocklist(ipBlocklist)
                            .setDatacenter(datacenter)
                    );
            final SearchEventsResponse events = apiResponse.getData();
            final int code = apiResponse.getStatusCode();
            final MusicianResponse response = new MusicianResponse(code, events, events);
            return ResponseEntity.ok(response);
        } catch (ApiException e) {
            final MusicianResponse response = new MusicianResponse(e);
            return ResponseEntity.ok(response);
        }
    }

    @GetMapping("/getEvents")
    public ResponseEntity<MusicianResponse> getEvents(
            @RequestParam(required = false, value = "") String apiKey,
            @RequestParam(required = false, value = "") String region,
            @RequestParam(required = false, value = "") String requestId
    ) {
        ApiClient client = Configuration.getDefaultApiClient(apiKey, Utils.getRegion(region));
        FingerprintApi api = new FingerprintApi(client);
        try {
            final ApiResponse<EventsGetResponse> apiResponse = api.getEventWithHttpInfo(requestId);
            final EventsGetResponse event = apiResponse.getData();
            final int code = apiResponse.getStatusCode();
            final MusicianResponse response = new MusicianResponse(code, event, event);
            return ResponseEntity.ok(response);
        } catch (ApiException e) {
            final MusicianResponse response = new MusicianResponse(e);
            return ResponseEntity.ok(response);
        }
    }

    @GetMapping("/updateEvent")
    public ResponseEntity<MusicianResponse> updateEvent(
            @RequestParam(required = false, value = "") String apiKey,
            @RequestParam(required = false, value = "") String region,
            @RequestParam(required = false, value = "") String requestId,
            @RequestParam(required = false) String linkedId,
            @RequestParam(required = false) String tag,
            @RequestParam(required = false) Boolean suspect
    ) {
        ApiClient client = Configuration.getDefaultApiClient(apiKey, Utils.getRegion(region));
        FingerprintApi api = new FingerprintApi(client);
        try {
            final EventsUpdateRequest eventsUpdateRequest = new EventsUpdateRequest();
            eventsUpdateRequest.setLinkedId(linkedId);
            eventsUpdateRequest.setSuspect(suspect);
            if (tag != null) {
                try {
                    Map<String, Object> parsedTag = objectMapper.readValue(tag, new TypeReference<Map<String, Object>>() {
                    });
                    eventsUpdateRequest.setTag(parsedTag);
                } catch (JsonProcessingException e) {
                }
            }

            final ApiResponse<Void> apiResponse = api.updateEventWithHttpInfo(requestId, eventsUpdateRequest);
            final int code = apiResponse.getStatusCode();
            final MusicianResponse response = new MusicianResponse(code, apiResponse.getData(), apiResponse.getData());
            return ResponseEntity.ok(response);
        } catch (ApiException e) {
            final MusicianResponse response = new MusicianResponse(e);
            return ResponseEntity.ok(response);
        }
    }
}

File: projects/dotnet-sdk/Controllers/EventsController.cs
using Microsoft.AspNetCore.Mvc;
using FingerprintPro.ServerSdk.Api;
using FingerprintPro.ServerSdk.Client;
using FingerprintPro.ServerSdk.Model;
using dotnet_sdk.Models;
using FingerprintPro.ServerSdk.Json;

namespace dotnet_sdk.Controllers
{
    [ApiController]
    [Route("")]
    public class EventsController : ControllerBase
    {
        [HttpGet("/getEvents")]
        public async Task<IActionResult> GetEvents(
            [FromQuery] string? apiKey,
            [FromQuery] string? region,
            [FromQuery] string? requestId)
        {
            try
            {
                var parsedRegion = Utils.GetRegion(region);
                var configuration = new Configuration(apiKey)
                {
                    Region = parsedRegion
                };
                var api = new FingerprintApi(configuration);

                var apiResponse = api.GetEventWithHttpInfo(requestId);
                var eventResponse = apiResponse.Response;
                var rawResponse = await eventResponse.Content.ReadAsStringAsync();

                var response = new MusicianResponse<EventsGetResponse>(eventResponse.StatusCode, rawResponse, apiResponse.Data);
                return Ok(response);
            }
            catch (Exception e) {
                return Ok(Utils.ProcessException(e));
            }
        }

        [HttpGet("/searchEvents")]
        public async Task<IActionResult> SearchEvents(
            [FromQuery] string? apiKey,
            [FromQuery] string? region,
            [FromQuery] int? limit,
            [FromQuery] string? paginationKey,
            [FromQuery] string? visitorId,
            [FromQuery] string? bot,
            [FromQuery] string? ipAddress,
            [FromQuery] string? linkedId,
            [FromQuery] long? start,
            [FromQuery] long? end,
            [FromQuery] bool? reverse,
            [FromQuery] bool? suspect,
            [FromQuery] bool? vpn,
            [FromQuery] bool? virtualMachine,
            [FromQuery] bool? tampering,
            [FromQuery] bool? antiDetectBrowser,
            [FromQuery] bool? incognito,
            [FromQuery] bool? privacySettings,
            [FromQuery] bool? jailbroken,
            [FromQuery] bool? frida,
            [FromQuery] bool? factoryReset,
            [FromQuery] bool? clonedApp,
            [FromQuery] bool? emulator,
            [FromQuery] bool? rootApps,
            [FromQuery] float? minSuspectScore,
            [FromQuery] bool? ipBlocklist,
            [FromQuery] bool? datacenter
            )
        {
            try
            {
                var parsedRegion = Utils.GetRegion(region);
                var configuration = new Configuration(apiKey)
                {
                    Region = parsedRegion
                };
                var api = new FingerprintApi(configuration);

                var apiResponse = api.SearchEventsWithHttpInfo(limit, paginationKey: paginationKey, visitorId:visitorId, bot:bot, ipAddress:ipAddress, linkedId:linkedId,
                                                               start:start, end:end, reverse:reverse, suspect:suspect, vpn:vpn, virtualMachine:virtualMachine,
                                                               tampering:tampering, antiDetectBrowser:antiDetectBrowser, incognito:incognito, privacySettings:privacySettings,
                                                               jailbroken:jailbroken, frida:frida, factoryReset:factoryReset, clonedApp:clonedApp, emulator:emulator,
                                                               rootApps:rootApps, minSuspectScore:minSuspectScore, ipBlocklist:ipBlocklist, datacenter:datacenter);
                var eventResponse = apiResponse.Response;
                var rawResponse = await eventResponse.Content.ReadAsStringAsync();

                var response = new MusicianResponse<SearchEventsResponse>(eventResponse.StatusCode, rawResponse, apiResponse.Data);
                return Ok(response);
            }
            catch (Exception e) {
                return Ok(Utils.ProcessException(e));
            }
        }

        [HttpGet("/updateEvent")]
        public async Task<IActionResult> UpdateEvent(
            [FromQuery] string? apiKey,
            [FromQuery] string? region,
            [FromQuery] string? requestId,
            [FromQuery] string? linkedId,
            [FromQuery] string? tag,
            [FromQuery] Boolean? suspect
            )
        {
            try
            {
                var parsedRegion = Utils.GetRegion(region);
                var configuration = new Configuration(apiKey)
                {
                    Region = parsedRegion
                };
                var api = new FingerprintApi(configuration);

                var deserializedTag = tag != null ? JsonUtils.Deserialize<Tag>(tag) : null;

                var updateEventRequest = new EventsUpdateRequest(linkedId, deserializedTag, suspect);

                var apiResponse = api.UpdateEventWithHttpInfo(updateEventRequest, requestId);
                var eventResponse = apiResponse.Response;
                var rawResponse = await eventResponse.Content.ReadAsStringAsync();

                var response = new MusicianResponse<object>(eventResponse.StatusCode, rawResponse, apiResponse.Data);
                return Ok(response);
            }
            catch (Exception e) {
                return Ok(Utils.ProcessException(e));
            }
        }
    }
}
